### PR TITLE
add match lstm model

### PR DESCRIPTION
**Pull request by faneshion**
#436

### CODE REVIEW

File: matchzoo/models/match_lstm.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+"""Match LSTM model."""
+import keras
+import keras.backend as K
+from matchzoo import engine, preprocessors
+
+
+class MLSTMModel(engine.BaseModel):
```
Comment (uduse) [inline matchzoo/models/match_lstm.py]: Match your class name with your file name.

File: matchzoo/models/match_lstm.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        >>> model = MLSTMModel()
+        >>> model.guess_and_fill_missing_params(verbose=0)
+        >>> model.build()
+
+    """
+
+    @classmethod
+    def get_default_preprocessor(cls):
+        """Get default preprocessor."""
+        return preprocessors.NaivePreprocessor()
```
Comment (uduse) [inline matchzoo/models/match_lstm.py]: `BaseModel.get_default_preprocessor` already returns `NaivePreprocessor`. However, if this is just a placeholder for future development, then it's fine.

File: matchzoo/models/match_lstm.py (hunk 1 above)
Comment (lixinsu) [reply matchzoo/models/match_lstm.py]: I have changed the model name

File: matchzoo/models/match_lstm.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    """
+
+    @classmethod
+    def get_default_params(cls):
+        """Get default parameters."""
+        params = super().get_default_params(with_embedding=True)
+        params.add(engine.Param(
+            'rnn_hidden_size', 256,
+            hyper_space=engine.hyper_spaces.quniform(low=128, high=384, q=128)
```
Comment (uduse) [inline matchzoo/models/match_lstm.py]: `q` is step size. Should be 1 here.

File: matchzoo/models/match_lstm.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            'fc_hidden_size', 200,
+            hyper_space=engine.hyper_spaces.quniform(
+                low=100, high=300, q=100)
+        ))
+        return params
+
+    def build(self):
+        """Build model."""
+        query, doc = self._make_inputs()
+        self.query_len = query.shape[1]
```
Comment (uduse) [inline matchzoo/models/match_lstm.py]: Do not store things in `self`. Change the signature of your attention method to take those as parameters.

File: matchzoo/models/match_lstm.py (hunk 3 above)
Comment (lixinsu) [reply matchzoo/models/match_lstm.py]: I choose a smaller step value

File: matchzoo/models/match_lstm.py (hunk 4 above)
Comment (lixinsu) [reply matchzoo/models/match_lstm.py]: When a tenor is passed through  LSTM layer in Keras, the seq_length dimension of the output tensor is None. And the attention function only can receive tensor as its parameter, so I change the attention as an inner function of build to get get the length information.

File: matchzoo/models/match_lstm.py (hunk 3 above)
Comment (lixinsu) [reply matchzoo/models/match_lstm.py]: I think it is not necessary to search every possible value for the hidden size

Comment (uduse) [reply matchzoo/models/match_lstm.py]: @lixinsu It's not a grid search. `q` merely affects the granularity of sampled values.

File: matchzoo/models/match_lstm.py (hunk 4 above)
Comment (uduse) [reply matchzoo/models/match_lstm.py]: @lixinsu Yep that is a better way. An alternative is to make it a `classmethod`.

**Pull request merged**